SuperChamps/contracts/Metagame/SCMetagamePool.sol

```diff
@@ -0,0 +1,135 @@
+
+    /// @param permissions_ Address of the protocol permissions registry. Must conform to IPermissionsManager
+    constructor(address permissions_, address token_) SCPermissionedAccess(permissions_) {
+        token = IERC20(token_);
+    }
+
+    /// @notice Stake tokens directly
+    /// @param amount_ Quantity of tokens to stake
+    function stake(uint256 amount_) external {
+        bool success = token.transferFrom(msg.sender, address(this), amount_);
+        if(!success) {
+            revert UnableToTransferTokens(msg.sender, amount_);
+        }
+        _stake(msg.sender, amount_);
+    }
+
+    /// @notice Stake tokens for another address
+    /// @param staker_ The address to stake tokens for
+    /// @param amount_ Quantity of tokens to stake
+    /// @dev Tokens are transferred from msg.sender and credited to staker_
+    function stakeFor(address staker_, uint256 amount_) external {
+        bool success = token.transferFrom(msg.sender, address(this), amount_);
+        if(!success) {
+            revert UnableToTransferTokens(msg.sender, amount_);
+        }
+        _stake(staker_, amount_);
+    }
+
+    /// @notice Approve an address to spend staked tokens
+    /// @param spender_ The address to approve
+    /// @param amount_ Quantity of tokens to approve spending
+    /// @dev This mirrors approve(...) defined in ERC20 standard
+    function approve(address spender_, uint256 amount_) external {
+        _user_to_approved_spend[msg.sender][spender_] += amount_;
+    }
+
+    /// @notice Spend tokens from a stakers staked tokens
+    /// @param amount_ Quantity of tokens to unstake/spend
+    /// @param staker_ The address of the staker
+    /// @param receiver_ The address of the reciever of the spent tokens
+    /// @dev This does NOT trigger an Unstake event, allowing stakers to spend tokens without penalty
+    /// @dev This is expected to be called from the spending contract, which must be first permissioned through a direct call to approve(...)
+    function spend(uint256 amount_, address staker_, address receiver_) external {
+        if(staker_ != msg.sender) {
+            _user_to_approved_spend[staker_][msg.sender] -= amount_;
+        }
+
+        uint256 _balance = _unstake(amount_, staker_, receiver_);
+        emit SpendFromStake(staker_, amount_, _balance);
+    }
+
+    /// @notice Unstake tokens and transfer to staker
+    /// @param amount_ Quantity of tokens to unstake
+    /// @dev This triggers an Unstake event which can cause penalties in the metagame
+    function unstake(uint256 amount_) external {
+        uint256 _balance = _unstake(amount_, msg.sender, msg.sender);
+        emit Unstake(msg.sender, amount_, _balance);
+    }
+
+    /// @notice Returns the list of all of the users staking checkpoint timestamps
+    /// @param staker_ Address of the staker
+    function checkpoint_timestamps(address staker_) public view returns (uint256[] memory) {
+        return _user_checkpoints[staker_];
```

Comment:
@figs999  will it cause issue if returned array is too long like 5-10k items in array?